### PR TITLE
Migrating java-util from Metamarkets.

### PR DESCRIPTION
**Pull request by akashdw**
PR to migrate metamx/java-util. This migration will be helpful to deal problems around having separate PRs and release cycle of Druid and java-util.

metamx/java-util is still coming in because of transitive dependencies(emitter and server-metrics), will create separate PR's to move metamx/emitter, metamx/server-metrics and metamx/http-client.

Note: External extensions might have to change.

### CODE REVIEW

**Comment by b-slim** [comment]
@akashdw build is failing mostly because of code style can you please fix that ? use the intellij template it will fix it thought.

**Comment by akashdw** [comment]
Thanks @b-slim, I'm using eclipse and using eclipse-formatting.xml for formatting.

**Comment by akashdw** [comment]
Thanks @b-slim, Most of the formatting errors are coming from java-util. I'm using eclipse-formatting.xml, migrating to intellij  to resolve formatting error.

**Comment by b-slim** [comment]
FYI closing/opening the PR will not fix this.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check (validate) on project java-util: Failed during checkstyle execution: There are 82 errors reported by Checkstyle 6.11.2 with codestyle/checkstyle.xml ruleset. -> [Help 1]

[ERROR] 
```

**Comment by akashdw** [comment]
@b-slim Closing pr was not intentional and I hit the `Close and comment` button while replying.

File: NOTICE [hunk 1]
```diff
@@ -22,3 +22,9 @@ This product contains variable length long deserialization code adapted from Apa
     * https://github.com/apache/lucene-solr/blob/master/lucene/LICENSE.txt (Apache License, Version 2.0)
   * HOMEPAGE:
     * https://github.com/apache/lucene-solr
+
+This product contains a modified version of Metamarkets java-util library
+  * LICENSE:
+    * https://github.com/metamx/java-util/blob/master/LICENSE (Apache License, Version 2.0)
+  * HOMEPAGE:
+    * https://github.com/metamx/java-util
```
Comment (drcrallen) [inline NOTICE]: Can you put a link in here to the commit tag you forked from?

**Comment by leventov** [comment]
In IntelliJ, it's easy to fix many formatting errors at once by Ctrl/Cmd+Shift+A, "run inspection by name", then find a relevant inspection like "unused import" or "if-else without braces", run the inspection. It will give a list of inspection violations. Then click the "resolve all" button.

File: NOTICE (hunk 1 above)
Comment (akashdw) [inline NOTICE]: done

**Comment by nishantmonu51** [comment]
still some more checkstyle errors related to long literal suffix - 
[LongLiteralLowerCaseSuffix] Prefer 'L' to 'l' for the suffix to long literals

File: NOTICE (hunk 1 above)
Comment (leventov) [inline NOTICE]: @akashdw could you please also cherry-pick this commit: https://github.com/metamx/java-util/commit/826021f37717b6638ce5fee9543456cce4678484?

**Comment by leventov** [comment]
@nishantmonu51 "LongLiteralLowerCaseSuffix" doesn't seem to be a Checkstyle rule. There are only a few enabled rules now: https://github.com/druid-io/druid/blob/master/codestyle/checkstyle.xml

**Comment by nishantmonu51** [comment]
@leventov: I just looked at the errors from the travis logs here - 
https://travis-ci.org/druid-io/druid/jobs/168497040

```
/home/travis/build/druid-io/druid/java-util/src/main/java/io/druid/java/util/common/FileUtils.java:89: error: [LongLiteralLowerCaseSuffix] Prefer 'L' to 'l' for the suffix to long literals
    private long size = 0l;
                        ^
    (see http://errorprone.info/bugpattern/LongLiteralLowerCaseSuffix)
  Did you mean 'private long size = 0L;'?
/home/travis/build/druid-io/druid/java-util/src/main/java/io/druid/java/util/common/FileUtils.java:117: error: [LongLiteralLowerCaseSuffix] Prefer 'L' to 'l' for the suffix to long literals
      long size = 0l;
                  ^
    (see http://errorprone.info/bugpattern/LongLiteralLowerCaseSuffix)
  Did you mean 'long size = 0L;'?
/home/travis/build/druid-io/druid/java-util/src/main/java/io/druid/java/util/common/StreamUtils.java:140: error: [LongLiteralLowerCaseSuffix] Prefer 'L' to 'l' for the suffix to long literals
    long size = 0l;
                ^
    (see http://errorprone.info/bugpattern/LongLiteralLowerCaseSuffix)
  Did you mean 'long size = 0L;'?
```

File: NOTICE (hunk 1 above)
Comment (akashdw) [inline NOTICE]: Done  👍 .

**Comment by akashdw** [comment]
Thanks @b-slim, @nishantmonu51 and @leventov, updated the pr with formatting fix and cherry picked  metamx/java-util@826021f.

**Comment by drcrallen** [comment]
@akashdw you don't need to force push the branch anymore, git will do all the rebasing for us now.

**Comment by drcrallen** [comment]
The copyright in the POM here is consistent with other Druid poms but is not consistent with the java copyright. I'm inclined to say the copyright should match the java one and to do another PR to fix the rest of the files.

**Comment by drcrallen** [comment]
can you remove the commented out dependency on java-util test scope in the mater pom.xml?

**Comment by drcrallen** [comment]
`processing/src/main/java/io/druid/query/RetryQueryRunner.java` has a code change that should probably not be there.

**Comment by drcrallen** [comment]
same with `server/src/main/java/io/druid/guice/http/HttpClientModule.java`

**Comment by drcrallen** [comment]
same with `server/src/main/java/io/druid/server/initialization/HttpEmitterModule.java`

**Comment by drcrallen** [comment]
and `server/src/test/java/io/druid/server/initialization/BaseJettyTest.java`

I'm looking at a diff between `b113a34355a6fdbf115b737299aced1815138cde` and head of this branch, which should be only the changes suggested for the migration

**Comment by akashdw** [comment]
@drcrallen I just rebased locally and can see those file changes.

**Comment by drcrallen** [comment]
Can you make a `Lifecycle` static method with the following signature?

```java 
public static com.metamx.common.lifecycle.Lifecycle asMmxLifecycle(Lifecycle lifecycle)
```

Or maybe just a method as

```java 
public com.metamx.common.lifecycle.Lifecycle asMmxLifecycle()
```

**Comment by drcrallen** [comment]
Actually, that probably works best as a `LifecycleUtils` static method

**Comment by drcrallen** [comment]
It won't let me do a formal review thing for some reason. :-P I finished looking through the code and I think only `RetryQueryRunner` looks unintentional.

Please also remove the commented out XML in the master pom related to java-util-test

**Comment by drcrallen** [comment]
Thanks for going through with this!

**Comment by akashdw** [comment]
@drcrallen Thanks for reviewing this PR 👍. I uploaded new patch set for pom, copyright, `RetryQueryRunner` and  `LifecycleUtils` changes.

**Comment by drcrallen** [comment]
:+1:

**Comment by cheddar** [comment]
👍

**Pull request merged**